server/midddleware/forwarder.go

```diff
@@ -86,9 +120,92 @@ func forwarderUnaryServerInterceptor() grpc.UnaryServerInterceptor {
 		txCtx := api.GetTransaction(ctx, req.(proto.Message))
 
 		if txCtx != nil && txCtx.GetOrigin() != types.MyOrigin {
-			return forwardRequest(ctx, txCtx, info.FullMethod, req.(proto.Message))
+			iface, err := forwardRequest(ctx, txCtx, info.FullMethod, req.(proto.Message))
+			log.Err(err).Str("method", info.FullMethod).Msg("forwarded request")
+			return iface, err
 		}
 
 		return handler(ctx, req)
 	}
 }
+
+func proxyHandler(_ interface{}, serverStream grpc.ServerStream) error {
+	fullMethodName, ok := grpc.MethodFromServerStream(serverStream)
+	if !ok {
+		return api.Errorf(api.Code_INTERNAL, "failed to determine method name")
+	}
+
+	outgoingCtx, backendConn, req, resp, err := proxyDirector(serverStream.Context(), fullMethodName)
+	if err != nil {
+		return err
+	}
+
+	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
+	defer clientCancel()
+
+	clientStream, err := grpc.NewClientStream(clientCtx, &grpc.StreamDesc{ServerStreams: true, ClientStreams: true}, backendConn, fullMethodName)
+	if err != nil {
+		return err
+	}
+
+	downErr := <-forwardDownstream(serverStream, req, clientStream)
+	_ = clientStream.CloseSend()
+
+	upErr := <-forwardUpstream(clientStream, resp, serverStream)
```

Comment:
Line 151(downErr := <- ..) is blocking? Which means if there is an upErr we won't see it because thread will be blocked at downErr till there is an error from downStream?

Reply:
so you think better to have select here?

Reply:
Yeah, then with either error we can return.

Reply:
Actually this is totally wrong. Let me think how to manage this.